src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java

```diff
@@ -0,0 +1,354 @@
+            validateGitlabResponse(httpResponse, 201, "Comment posted");
+        }
+    }
+
+    private void postStatus(String statusPostUrl, Map<String, String> headers, AnalysisDetails analysis, String coverage, boolean sendRequest) throws IOException{
+        //See https://docs.gitlab.com/ee/api/commits.html#post-the-build-status-to-a-commit
+        statusPostUrl += "?name=SonarQube";
+        String status = (analysis.getQualityGateStatus() == QualityGate.Status.OK ? "success" : "failed");
+        statusPostUrl += "&state=" + status;
+        statusPostUrl += "&target_url=" + URLEncoder.encode(String.format("%s/dashboard?id=%s&pullRequest=%s", server.getPublicRootUrl(),
+                URLEncoder.encode(analysis.getAnalysisProjectKey(),
+                        StandardCharsets.UTF_8.name()), URLEncoder
+                        .encode(analysis.getBranchName(),
+                                StandardCharsets.UTF_8.name())), StandardCharsets.UTF_8.name());
+        statusPostUrl+="&description="+URLEncoder.encode("SonarQube Status", StandardCharsets.UTF_8.name());
+        statusPostUrl+="&coverage="+coverage;
+        //TODO: add pipelineId if we have it
+
+        HttpPost httpPost = new HttpPost(statusPostUrl);
+        for (Map.Entry<String, String> entry : headers.entrySet()) {
+            httpPost.addHeader(entry.getKey(), entry.getValue());
+        }
+        if (sendRequest) {
+            HttpResponse httpResponse = HttpClients.createDefault().execute(httpPost);
+            if (null != httpResponse && httpResponse.toString().contains("Cannot transition status")) {
+                // Workaround for https://gitlab.com/gitlab-org/gitlab-ce/issues/25807
+                LOGGER.debug("Transition status is already {}", status);
+            } else {
+                validateGitlabResponse(httpResponse, 201, "Comment posted");
+            }
+        }
+    }
+
+    private void validateGitlabResponse(HttpResponse httpResponse, int expectedStatus, String successLogMessage) throws IOException {
+        if (null != httpResponse && httpResponse.getStatusLine().getStatusCode() != expectedStatus) {
+            LOGGER.error(httpResponse.toString());
+            LOGGER.error(EntityUtils.toString(httpResponse.getEntity(), StandardCharsets.UTF_8));
+            throw new IllegalStateException("An error was returned in the response from the Gitlab API. See the previous log messages for details");
+        } else if (null != httpResponse) {
+            LOGGER.debug(httpResponse.toString());
+            LOGGER.info(successLogMessage);
+        }
+    }
+
+    private static String getMandatoryProperty(String propertyName, Configuration configuration) {
+        return configuration.get(propertyName).orElseThrow(() -> new IllegalStateException(
+                String.format("%s must be specified in the project configuration", propertyName)));
+    }
+
+    private static String getNextUrl(HttpResponse httpResponse) {
```

Comment:
Nitpick: should this not return `Optional<String>` rather than using `null` and requiring calling code to perform a null check?